### PR TITLE
[Enhancement](runtime-filter) enlarge runtime filter in predicate threshold

### PR DESCRIPTION
**Pull request by BiteTheDDDDt**
# Proposed changes
TPCH-Q17:
before:
图片
图片

after:
图片
图片

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 37.68 seconds
 load time: 564 seconds
 storage size: 17154887239 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221022232620_clickbench_pr_32791.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.67 seconds
 load time: 560 seconds
 storage size: 17154655391 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221026101841_clickbench_pr_34284.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.64 seconds
 load time: 564 seconds
 storage size: 17154644814 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221026182535_clickbench_pr_34289.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.65 seconds
 load time: 564 seconds
 storage size: 17154644810 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221026200038_clickbench_pr_34322.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 39.15 seconds
 load time: 559 seconds
 storage size: 17154655346 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221027085925_clickbench_pr_34676.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.67 seconds
 load time: 609 seconds
 storage size: 17154810701 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221027205246_clickbench_pr_34824.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 39.14 seconds
 load time: 598 seconds
 storage size: 17154821103 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028044531_clickbench_pr_35112.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 39.41 seconds
 load time: 571 seconds
 storage size: 17154821270 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028141904_clickbench_pr_35169.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.4 seconds
 load time: 569 seconds
 storage size: 17154644748 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028162723_clickbench_pr_35231.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 39.09 seconds
 load time: 572 seconds
 storage size: 17154865254 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028190121_clickbench_pr_35328.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.89 seconds
 load time: 568 seconds
 storage size: 17154644783 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028113112_clickbench_pr_35349.html

File: be/src/exprs/in_predicate.cpp [hunk 1]
```diff
@@ -35,13 +33,17 @@ InPredicate::InPredicate(const TExprNode& node)
           _null_in_set(false),
           _hybrid_set() {}
 
-InPredicate::~InPredicate() {}
+InPredicate::~InPredicate() {
+    if (_should_delete) {
```
Comment (HappenLee) [inline be/src/exprs/in_predicate.cpp:37]: really we need `should_delete` ? use  _hybrid_set == nullptr or just use unique_ptr?

File: be/src/olap/in_list_predicate.h [hunk 2]
```diff
@@ -101,24 +102,78 @@ class InListPredicateBase : public ColumnPredicate {
...
+            } else if constexpr (Type == TYPE_DATE) {
+                for (auto& value : *values) {
+                    _values->insert(value.to_olap_date());
+                }
+            } else if constexpr (Type == TYPE_DATETIME) {
+                for (auto& value : *values) {
+                    _values->insert(value.to_olap_datetime());
+                }
+            } else {
+                for (auto& value : *values) {
```
Comment (HappenLee) [inline be/src/olap/in_list_predicate.h:147]: only ` is_string_type(Type) || Type == TYPE_DECIMALV2 || is_date_type(Type)` which case will be here ? maybe need DCHECK?

File: be/src/olap/in_list_predicate.h [hunk 3]
```diff
@@ -101,24 +102,78 @@ class InListPredicateBase : public ColumnPredicate {
...
+        if constexpr (is_string_type(Type) || Type == TYPE_DECIMALV2 || is_date_type(Type)) {
+            _values = new phmap::flat_hash_set<T>();
+            auto values = ((HybridSetType*)hybrid_set.get())->get_inner_set();
+
+            if constexpr (is_string_type(Type)) {
+                for (auto& value : *values) {
+                    StringValue sv = {value.data(), int(value.size())};
+                    if constexpr (Type == TYPE_CHAR) {
+                        _temp_datas.push_back("");
+                        _temp_datas.back().resize(std::max(char_length, value.size()));
```
Comment (HappenLee) [inline be/src/olap/in_list_predicate.h:128]: resize can make sure the new len is filled by "\0" ?

Comment (BiteTheDDDDt) [reply be/src/olap/in_list_predicate.h:128]: https://cplusplus.com/reference/string/string/resize/ 
`they are value-initialized characters.`  I think it means `char()`. 
In other words, value-initialized characters is '\0'.

File: be/src/exprs/in_predicate.cpp (hunk 1 above)
Comment (BiteTheDDDDt) [reply be/src/exprs/in_predicate.cpp:37]: InPredicate have another prepare function `InPredicate::prepare(RuntimeState* state, HybridSetBase* hset)`, when this prepare be used, we should do not delete `_hybrid_set`

File: be/src/vec/exprs/vdirect_in_predicate.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                _filter->find_fixed_len(argument_column->get_raw_data().data, nullptr, sz, ptr);
+            }
+        } else {
+            for (size_t i = 0; i < sz; i++) {
+                ptr[i] = _filter->find(
+                        reinterpret_cast<const void*>(argument_column->get_data_at(i).data));
+            }
+        }
+
+        if (_data_type->is_nullable()) {
```
Comment (HappenLee) [inline be/src/vec/exprs/vdirect_in_predicate.h:72]: this function only use by rf ? maybe do not need nullable column of result

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**